src/common.cpp

```diff
@@ -1,10 +1,93 @@
 #include "common.hh"
+#include "zarr.hh"
 
 #include "platform.h"
 
 #include <cmath>
+#include <thread>
 
 namespace common = acquire::sink::zarr::common;
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : contexts_{ n_threads }
+  , jobs_mutex_{ std::make_shared<std::mutex>() }
+  , error_handler_{ err }
+{
+    if (n_threads == 0) {
+        throw std::runtime_error("Cannot create thread pool with 0 threads.");
+    }
+
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = false;
+        ctx.thread = std::thread([this, ctx_ = &ctx] { thread_worker_(ctx_); });
+        ctx.mutex = jobs_mutex_;
+    }
+}
+
+common::ThreadPool::~ThreadPool() noexcept
+{
+    await_stop();
+}
+
+void
+common::ThreadPool::push_to_job_queue(JobT&& job)
+{
+    std::scoped_lock lock(*jobs_mutex_);
+    jobs_.push(std::move(job));
+
+    for (auto& ctx : contexts_) {
```

Comment:
Optional: this `for` loop pattern with `notify_one` seems a bit weird to me, as we know one job has been added here, but we can potentially notify all the threads in the pool. I think a viable alternative solution is only have one condition variable that's associated with the `ThreadPool`, wait on that in `thread_worker_` which should have a live `ThreadPool` instance, then just call `notify_one` here so at most one thread is notified.

This is optional because notifying all of them works, but only the first waiting one will win, so it's sub-optimal.

Reply:
No, this makes a lot more sense than what I have and is much simpler.